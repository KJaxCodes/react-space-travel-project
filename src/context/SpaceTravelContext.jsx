import React from 'react';
import { createContext, useContext, useState } from 'react';
import SpaceTravelApi from '../services/SpaceTravelApi';

const SpaceTravelContext = createContext();

export const useSpaceTravel = () => useContext(SpaceTravelContext);

export const SpaceTravelProvider = ({ children }) => {
    const [spacecrafts, setSpacecrafts] = useState([]);
    const [planets, setPlanets] = useState([]);
    const [isLoading, setIsLoading] = useState(false);


    //Actions

    //Decommission spacecraft by ID
    const decommissionSpacecraftById = async (id) => {

        try {
            setIsLoading(true);
            const response = await SpaceTravelApi.destroySpacecraftById({ id: id });
            if (response.isError) {
                throw new Error("Failed to decommission.");
            }
            setSpacecrafts((prev) => prev.filter((craft) => craft.id !== id));
        } catch (err) {
            console.error("Decommission failure", err.message);
        } finally {
            console.log("This will always run");
            setIsLoading(false);
            console.log("And we are done loading");
        }
    };

    //Assign or reassign spacecraft to planet
    const assignSpacecraftToPlanet = async ({ spacecraftId, targetPlanetId }) => {
        try {
            const response = await SpaceTravelApi.sendSpacecraftToPlanet({
                spacecraftId, targetPlanetId
            });
            if (!response.isError) {
                setSpacecrafts((prev) =>
                    prev.map((craft) =>
                        craft.id === spacecraftId ? { ...craft, currentLocation: targetPlanetId } : craft
                    )
                );
            }
        } catch (err) {
            console.error("Failed to assign spacecraft", err.message);
        }
    }

    // Load initial spacecrafts and planets

    const loadInitialData = async () => {
        try {
            setIsLoading(true);

            const spacecraftRes = await SpaceTravelApi.getSpacecrafts();
            const planetRes = await SpaceTravelApi.getPlanets();

            if (spacecraftRes.isError) {
                console.log("loadInitialData Error: Spacecraft");
                console.debug(spacecraftRes);
            }
            if (planetRes.isError) {
                console.log("loadInitialData Error: Planet");
                console.debug(spacecraftRes);
            }

            setSpacecrafts(spacecraftRes.data);
            setPlanets(planetRes.data);

        } catch (err) {
            console.error("Failed to load initial data", err.message);
        } finally {
            setIsLoading(false);
        }
    }

    return (
        <SpaceTravelContext.Provider
            value={{
                spacecrafts,
                setSpacecrafts,
                planets,
                setPlanets,
                decommissionSpacecraftById,
                assignSpacecraftToPlanet,
                loadInitialData,
                isLoading,
                setIsLoading,
            }}
        >
            {children}
        </SpaceTravelContext.Provider>
    );
};