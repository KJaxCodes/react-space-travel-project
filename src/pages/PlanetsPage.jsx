import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { useSpaceTravel } from "../context/SpaceTravelContext";
import SpaceTravelApi from "../services/SpaceTravelApi"; // to get data from mock API
import SpacecraftSelector from "../components/SpacecraftSelector";

function PlanetsPage() {
  const {
    planets,
    setPlanets,
    spacecrafts,
    setSpacecrafts,
    decommissionSpacecraftById,
    spacecraftId,
    currentCraft,
    setCurrentCraft,

  } = useSpaceTravel();



  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  //event listeners and dispatchers

  const handleSetSelectedCraft = (craft) => {
    setCurrentCraft(craft);
  };

  const handleAssignCraftToPlanet = async (planetId) => {

    const craftId = currentCraft.id;
    try {
      const response = await SpaceTravelApi.sendSpacecraftToPlanet({ spacecraftId: craftId, targetPlanetId: planetId })
      console.log(response.data);
    } catch (e) {
      console.log(e);
    }
    console.log(currentCraft);
  };


  useEffect(() => {
    async function loadPlanets() {
      try {
        const response = await SpaceTravelApi.getPlanets();
        if (response.isError) {
          throw new Error("Failed to load planets.");
        }
        setPlanets(response.data);
      } catch (err) {
        setError(err.message);
      } finally {
        setIsLoading(false);
      }
    }

    async function loadSpacecrafts() {
      try {
        const response = await SpaceTravelApi.getSpacecrafts();
        if (response.isError) {
          throw new Error("Failed to load spacecraft.");
        }
        setSpacecrafts(response.data);
      } catch (err) {
        setError(err.message);
      } finally {
        setIsLoading(false);
      }
    }

    loadPlanets()
    loadSpacecrafts()

  }, []); // Empty dependency array so it runs only once after initial render

  useEffect(() => {
    if (planets.length > 0 && spacecrafts.length > 0) {
      console.log(planets);
      console.log(spacecrafts);
    }
  }, [planets, spacecrafts]);


  //TODO
  //fetch all current spacecraft available
  //set them to local state
  //then have a dropdown component which would render an N amount of spacecraft to select
  //upon selection: "Example: Brutus", the selection then filteres the fetched spacecraft by name
  //set the selected spacecraft to current -
  //have a button called assign
  //dispatches it by planet id and spacecraft id


  return (
    <div>
      <h2>Planets</h2>
      {isLoading && <p>Loading planets...âŒ›</p>}
      {error && <p style={{ color: "red" }}>Error: {error}</p>}
      {!isLoading && !error && (

        <ul>
          {planets.map((planet) => (
            <li key={planet.id}>
              <h4>{planet.name} </h4>
              <img src={planet.pictureUrl} alt={planet.name} style={{ width: "100px" }} />
              <p>Population: {planet.currentPopulation} </p>
              <p>Assigned Spacecraft:</p>
              <SpacecraftSelector spacecrafts={spacecrafts} currentCraft={currentCraft} handleSetSelectedCraft={handleSetSelectedCraft} />
              <button onClick={() => handleAssignCraftToPlanet(planet.id)}>Assign</button>

            </li>
          ))}
        </ul>
      )}
    </div>
  );
}

export default PlanetsPage;
