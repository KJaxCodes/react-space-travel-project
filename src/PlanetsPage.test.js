import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
//components
import PlanetsPage from "./pages/PlanetsPage";
import { SpaceTravelProvider } from "./context/SpaceTravelContext";
import { isTypedArray } from "util/types";
import SpaceTravelApi from "./services/SpaceTravelApi";
//interceptions - foundation for future end to end testing
import { mockGetPlanets, mockGetSpacecrafts } from "./test_helpers/interceptions";

jest.spyOn(SpaceTravelApi, "getPlanets").mockImplementation(mockGetPlanets);

jest.spyOn(SpaceTravelApi, "getSpacecrafts").mockImplementation(mockGetSpacecrafts);

describe("PlanetsPageTests", () => {

    // Rendering snapshot test
    test("Component renders correctly", async () => {

        const { asFragment } = render(
            <SpaceTravelProvider>
                <PlanetsPage />
            </SpaceTravelProvider>
        );

        expect(screen.getByText(/loading/i)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot("planets-page-loading");

        await waitFor(() => {
            const planetsComponent = screen.getAllByTestId("planets-page");
            expect(planetsComponent.length).toEqual(8);
        });

        expect(asFragment()).toMatchSnapshot("planets-page-display");

    });

});



