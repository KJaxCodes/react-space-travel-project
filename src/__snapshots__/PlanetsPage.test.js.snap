// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PlanetsPageTests Component renders correctly: planets-page-display 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Planets
    </h2>
    <ul>
      <li
        data-testid="planets-page"
      >
        <h4>
          Mercury 
        </h4>
        <img
          alt="Mercury"
          src="https://upload.wikimedia.org/wikipedia/commons/8/88/Reprocessed_Mariner_10_image_of_Mercury.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Venus 
        </h4>
        <img
          alt="Venus"
          src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/85/Venus_globe.jpg/800px-Venus_globe.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Earth 
        </h4>
        <img
          alt="Earth"
          src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/cb/The_Blue_Marble_%28remastered%29.jpg/800px-The_Blue_Marble_%28remastered%29.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 100000 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <ul>
          <li
            class="assigned"
          >
            Prispax
          </li>
        </ul>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Mars 
        </h4>
        <img
          alt="Mars"
          src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/02/OSIRIS_Mars_true_color.jpg/800px-OSIRIS_Mars_true_color.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Jupiter 
        </h4>
        <img
          alt="Jupiter"
          src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/9c/Jupiter%2C_image_taken_by_NASA%27s_Hubble_Space_Telescope%2C_June_2019.png/800px-Jupiter%2C_image_taken_by_NASA%27s_Hubble_Space_Telescope%2C_June_2019.png"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Saturn 
        </h4>
        <img
          alt="Saturn"
          src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ea/8423_20181_1saturn2016.jpg/1920px-8423_20181_1saturn2016.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Uranus 
        </h4>
        <img
          alt="Uranus"
          src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Uranus_as_seen_by_NASA%27s_Voyager_2_%28remastered%29_-_JPEG_converted.jpg/800px-Uranus_as_seen_by_NASA%27s_Voyager_2_%28remastered%29_-_JPEG_converted.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
      <li
        data-testid="planets-page"
      >
        <h4>
          Neptune 
        </h4>
        <img
          alt="Neptune"
          src="https://upload.wikimedia.org/wikipedia/commons/0/06/Neptune.jpg"
          style="width: 100px;"
        />
        <p>
          Population: 0 
        </p>
        <p>
          Assigned Spacecraft:
        </p>
        <p
          class="none"
        >
          No spacecraft assigned
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PlanetsPageTests Component renders correctly: planets-page-loading 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Planets
    </h2>
    <ul>
      <li>
        <p>
          ⏳Loading...⌛
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
